### PR TITLE
Convert UndoManager to Vala

### PR DESCRIPTION
**Pull request by tintou**

### CODE REVIEW

File: libcore/UndoManager.vala [hunk 1]
```diff
@@ -0,0 +1,993 @@
...
+            Marlin.UndoActionData? action = null;
+            lock (stack) {
+                action = stack_scroll_right ();
+                if (action != null) {
+                    action.locked = true;
+                }
+            }
+
+            do_menu_update ();
+            if (action != null) {
```
[the remark this answers is not in the document]
Comment (tintou) [reply libcore/UndoManager.vala]: Indeed

File: libcore/UndoManager.vala [hunk 2]
```diff
@@ -0,0 +1,993 @@
...
+        public signal void request_menu_update (UndoMenuData data);
+
+        public uint undo_levels { get; construct set; default = 10; }
+        public bool confirm_delete { get; construct set; default = false; }
+
+        private GLib.Queue<Marlin.UndoActionData> stack;
+        private uint index;
+        private bool undo_redo_flag;
+
+        public static uint64 get_file_modification_time (GLib.File file) {
```
[the remark this answers is not in the document]
Comment (tintou) [reply libcore/UndoManager.vala]: Moved

**Comment by tintou** [comment]
@jeremypw I've applied the required changes to the undo () and redo () functions, I'm unsure what to do with the possibly null undo_redo_data so I'm leaving it up to you to change them. I've removed the MenuAction struct and related functions, letting the window querying the has_undo/has_redo directly.

**Comment by jeremypw** [comment]
@tintou  I have pushed some fixes.  I just added null checks to silence the terminal warnings.  It would be possible to add another action_type  "UNDO-REDO" (or something) which is generated on undoing/redoing and which is never added to the stack but it did not seem worth the effort.

I also fixed a regression I noticed in undoing rename (and fixed a related existing issue).

**Comment by jeremypw** [comment]
@tintou I think I found a simpler (albeit slightly less efficient) way of dealing with (by avoiding) null action data.  Also fixed a small issue with redoing file/folder creation.

**Comment by jeremypw** [comment]
Fixing crash when trying to undo a cancelled move.

**Comment by jeremypw** [comment]
There still a question over what to do about partially completed file operations - at the moment an undo action is still added for these although this can cause problems e.g. when undoing a partially completed move/replace.  However, that issue exists also in master.

**Comment by jeremypw** [comment]
@tintou Are you OK with the changes I made?

**Comment by tintou** [comment]
@jeremypw Yes, they totally make sense

**Comment by jeremypw** [comment]
Seems to be a problem undoing "RestoreFromTrash" - it offers to permanently delete instead of moving back to trash.

**Comment by jeremypw** [comment]
@tintou OK, I am going to approve this now - it seems to work as well as master.  Please merge it if you  are satisfied there are no regressions.

**Pull request merged**